### ISSUE TITLE
`borrow_as_ptr` and `ptr_as_ptr` false positives in `windows::core::implement` macro invocation

### ISSUE THREAD

**Issue by PaulDance** [rust-lang/rust-clippy#15398]
### Summary

These lints trigger on the invocation of the external macro. As suggested in https://github.com/microsoft/windows-rs/issues/3690#issuecomment-3146215522, I'm also opening an issue here just in case. I haven't investigated any of `windows-rs`'s or Clippy's side of this yet, so there still remains to see where the responsibility should lie.

### Lint Name

`borrow_as_ptr` and `ptr_as_ptr`

### Reproducer

Using the following crate files:

```toml
# `Cargo.toml`
[package]
name = "tmp"
version = "0.1.0"
edition = "2024"

[dependencies]
windows = { version = "0.61.3", features = [
    "std",
    "Win32",
    "Win32_System",
    "Win32_System_Com",
]}
windows-core = "0.61.2"

[lints]
clippy.borrow_as_ptr = "warn"
clippy.ptr_as_ptr = "warn"
```

and

```rust
// `src/lib.rs`
use std::ffi::c_void;
use windows::Win32::System::Com::{IClassFactory, IClassFactory_Impl};
use windows::core::{BOOL, GUID, IUnknown, Ref, Result, implement};

#[implement(IClassFactory)]
struct A;

impl IClassFactory_Impl for A_Impl {
    fn CreateInstance(
        &self,
        _outer: Ref<'_, IUnknown>,
        _iid: *const GUID,
        _object: *mut *mut c_void,
    ) -> Result<()> {
        Ok(())
    }

    fn LockServer(&self, _lock: BOOL) -> Result<()> {
        Ok(())
    }
}
```

then

```console
$ cargo clippy --target x86_64-pc-windows-msvc
warning: `as` casting between raw pointers without changing their constness
 --> src/lib.rs:5:13
  |
5 | #[implement(IClassFactory)]
  |             ^^^^^^^^^^^^^ help: try `pointer::cast`, a safer alternative: `(IClassFactory as IClassFactory).cast::<IClassFactory>()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_as_ptr
  = note: requested on the command line with `-W clippy::ptr-as-ptr`

warning: borrow as raw pointer
 --> src/lib.rs:5:13
  |
5 | #[implement(IClassFactory)]
  |             ^^^^^^^^^^^^^ help: try: `&raw const IClassFactory`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#borrow_as_ptr
  = note: requested on the command line with `-W clippy::borrow-as-ptr`

warning: `as` casting between raw pointers without changing their constness
 --> src/lib.rs:5:13
  |
5 | #[implement(IClassFactory)]
  |             ^^^^^^^^^^^^^ help: try `pointer::cast`, a safer alternative: `IClassFactory.cast::<IClassFactory>()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_as_ptr

warning: `tmp` (lib) generated 3 warnings (run `cargo clippy --fix --lib -p tmp` to apply 3 suggestions)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.05s
```

The lints are triggered inside the code generated by the `implement` macro, while they shouldn't.

### Version

```text
rustc 1.88.0 (6b00bc388 2025-06-23)
binary: rustc
commit-hash: 6b00bc3880198600130e1cf62b8f8a93494488cc
commit-date: 2025-06-23
host: x86_64-unknown-linux-gnu
release: 1.88.0
LLVM version: 20.1.5
```

**Comment by ada4a** [comment]
This is indeed a false-positive in Clippy -- we generally avoid linting in proc-macros

**Comment by PaulDance** [comment]
What about https://github.com/microsoft/windows-rs/issues/3690#issuecomment-3181381319, though?

**Comment by ada4a** [comment]
That is the reason this issue happened. We usually check if a piece of code comes from an external macro, and if so avoid linting (since the user usually can't do anything about external code, as is your case). But since `implement` uses `quote_spanned!`, Clippy thinks that the code actually comes from the crate-internal code, and therefore lints. For those cases, we have the heavier artillery of `is_from_proc_macro`, which is more expensive but also recognizes more proc-macros. For a bit more context, see https://doc.rust-lang.org/clippy/development/macro_expansions.html#the-is_from_proc_macro-function

Does that answer your question?..

**Comment by PaulDance** [comment]
Yes, it definitely answers my question, thanks! My comment was more about whether the fix in Clippy was really necessary, since the fix in `windows-rs` will be attempted nevertheless.

**Comment by ada4a** [comment]
Ah, I see. I'm pretty sure the fix won't stop the lint from firing inside the macro _completely_, but only when Clippy runs _outside_ the `windows` crate (in your case, when you use the `implement` macro). Inside the `windows` crate though, the devs should be able to run Clippy and get the lint, as before

**Comment by PaulDance** [comment]
Got it, thanks!

**Comment by PaulDance** [comment]
👍

**Comment by PaulDance** [comment]
Just tested under `clippy 0.1.91 (425a9c0a0e 2025-08-17)` and it still triggers with the reproducer:

```console
$ cargo +nightly clippy --target x86_64-pc-windows-msvc
warning: `as` casting between raw pointers without changing their constness
 --> src/lib.rs:6:13
  |
6 | #[implement(IClassFactory)]
  |             ^^^^^^^^^^^^^ help: try `pointer::cast`, a safer alternative: `(IClassFactory as IClassFactory).cast::<IClassFactory>()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_as_ptr
  = note: requested on the command line with `-W clippy::ptr-as-ptr`

warning: borrow as raw pointer
 --> src/lib.rs:6:13
  |
6 | #[implement(IClassFactory)]
  |             ^^^^^^^^^^^^^ help: try: `&raw const IClassFactory`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#borrow_as_ptr
  = note: requested on the command line with `-W clippy::borrow-as-ptr`

warning: `as` casting between raw pointers without changing their constness
 --> src/lib.rs:6:13
  |
6 | #[implement(IClassFactory)]
  |             ^^^^^^^^^^^^^ help: try `pointer::cast`, a safer alternative: `IClassFactory.cast::<IClassFactory>()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_as_ptr
```

What does it mean with regards to the fix?

**Comment by ada4a** [comment]
Huh. I guess `is_from_proc_macro` isn't able to recognize this case for some reason? @Jarcho maybe you have any ideas?

**Comment by Jarcho** [comment]
The change hasn't been synced to the rust repo yet so nightly will still be broken. The next sync should be in a few days.

**Comment by PaulDance** [comment]
As I suspected, thanks. I'll update this then.
